File: app/src/main/java/com/krystianwsul/checkme/utils/TaskHierarchyContainer.java
package com.krystianwsul.checkme.utils;

import android.support.annotation.NonNull;

import com.google.common.collect.HashMultimap;
import com.krystianwsul.checkme.domainmodel.TaskHierarchy;

import junit.framework.Assert;

import java.util.HashMap;
import java.util.Set;

public class TaskHierarchyContainer<T, U extends TaskHierarchy> {
    @NonNull
    private final HashMap<T, U> mTaskHierarchiesById = new HashMap<>();

    @NonNull
    private final HashMultimap<TaskKey, U> mTaskHierarchiesByParent = HashMultimap.create();

    @NonNull
    private final HashMultimap<TaskKey, U> mTaskHierarchiesByChild = HashMultimap.create();

    public TaskHierarchyContainer() {

    }

    public void add(@NonNull T id, @NonNull U taskHierarchy) {
        Assert.assertTrue(!mTaskHierarchiesById.containsKey(id));

        mTaskHierarchiesById.put(id, taskHierarchy);
        Assert.assertTrue(mTaskHierarchiesByChild.put(taskHierarchy.getChildTaskKey(), taskHierarchy));
        Assert.assertTrue(mTaskHierarchiesByParent.put(taskHierarchy.getParentTaskKey(), taskHierarchy));
    }

    public void removeForce(@NonNull T id) {
        Assert.assertTrue(mTaskHierarchiesById.containsKey(id));

        U taskHierarchy = mTaskHierarchiesById.get(id);
        Assert.assertTrue(taskHierarchy != null);

        mTaskHierarchiesById.remove(id);

        TaskKey childTaskKey = taskHierarchy.getChildTaskKey();
        Assert.assertTrue(mTaskHierarchiesByChild.containsEntry(childTaskKey, taskHierarchy));

        Assert.assertTrue(mTaskHierarchiesByChild.remove(childTaskKey, taskHierarchy));

        TaskKey parentTaskKey = taskHierarchy.getParentTaskKey();
        Assert.assertTrue(mTaskHierarchiesByParent.containsEntry(parentTaskKey, taskHierarchy));

        Assert.assertTrue(mTaskHierarchiesByParent.remove(parentTaskKey, taskHierarchy));
    }

    @NonNull
    public Set<U> getByChildTaskKey(@NonNull TaskKey childTaskKey) {
        Set<U> set = mTaskHierarchiesByChild.get(childTaskKey);
        Assert.assertTrue(set != null);

        return set;
    }

    @NonNull
    public Set<U> getByParentTaskKey(@NonNull TaskKey parentTaskKey) {
        Set<U> set = mTaskHierarchiesByParent.get(parentTaskKey);
        Assert.assertTrue(set != null);

        return set;
    }

    @NonNull
    public U getById(@NonNull T id) {
        Assert.assertTrue(mTaskHierarchiesById.containsKey(id));

        return mTaskHierarchiesById.get(id);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/utils/TaskHierarchyContainer.kt
package com.krystianwsul.checkme.utils

import com.google.common.collect.HashMultimap
import com.krystianwsul.checkme.domainmodel.TaskHierarchy
import com.krystianwsul.common.utils.TaskKey

import java.util.*

class TaskHierarchyContainer<T, U : TaskHierarchy> {

    private val taskHierarchiesById = HashMap<T, U>()

    private val taskHierarchiesByParent = HashMultimap.create<TaskKey, U>()

    private val taskHierarchiesByChild = HashMultimap.create<TaskKey, U>()

    fun add(id: T, taskHierarchy: U) {
        check(!taskHierarchiesById.containsKey(id))

        taskHierarchiesById[id] = taskHierarchy
        check(taskHierarchiesByChild.put(taskHierarchy.childTaskKey, taskHierarchy))
        check(taskHierarchiesByParent.put(taskHierarchy.parentTaskKey, taskHierarchy))
    }

    fun removeForce(id: T) {
        check(taskHierarchiesById.containsKey(id))

        val taskHierarchy = taskHierarchiesById[id]!!

        taskHierarchiesById.remove(id)

        val childTaskKey = taskHierarchy.childTaskKey
        check(taskHierarchiesByChild.containsEntry(childTaskKey, taskHierarchy))

        check(taskHierarchiesByChild.remove(childTaskKey, taskHierarchy))

        val parentTaskKey = taskHierarchy.parentTaskKey
        check(taskHierarchiesByParent.containsEntry(parentTaskKey, taskHierarchy))

        check(taskHierarchiesByParent.remove(parentTaskKey, taskHierarchy))
    }

    fun getByChildTaskKey(childTaskKey: TaskKey): Set<U> = taskHierarchiesByChild.get(childTaskKey)!!

    fun getByParentTaskKey(parentTaskKey: TaskKey): Set<U> = taskHierarchiesByParent.get(parentTaskKey)!!

    fun getById(id: T) = taskHierarchiesById[id]!!

    val all: Collection<TaskHierarchy> get() = taskHierarchiesById.values
}
